src/PowerShellEditorServices.Host/CodeLens/PesterCodeLensProvider.cs

```diff
@@ -63,12 +74,43 @@ public PesterCodeLensProvider(EditorSession editorSession)
                     new ClientCommand(
                         "PowerShell.RunPesterTests",
                         "Debug tests",
-                        new object[] { scriptFile.ClientFilePath, true /* Run in debugger */, pesterSymbol.TestName })),
+                        new object[] { scriptFile.ClientFilePath, true /* Run in debugger */, pesterSymbol.TestName, describeBlockLineNumber })),
             };
 
             return codeLensResults;
         }
 
+        /// <summary>
+        /// Used to determine the value of <see cref="_pesterV4_6_0_OrHigherAvailable"/> as a background task.
+        /// </summary>
+        private bool DeterminePesterVersion()
+        {
+            var powerShell = new PSCommand()
+                .AddCommand("Get-Module")
+                .AddParameter("ListAvailable")
+                .AddParameter("Name", "Pester");
+
+            IEnumerable<PSObject> result = Task.Run(() =>
+            {
+                return _editorSession.PowerShellContext.ExecuteCommandAsync<PSObject>(powerShell);
+            }).Result;
+
+            if (result == null)
+            {
+                return false;
+            }
+
+            var minimumPesterVersionSupportingInlineInvocation = new Version(4, 6);
+            foreach (PSObject module in result)
```

Comment:
What I'm worried about here is... 

Lets look at the `PSModulePath`:
```
/Users/tyler/.local/share/powershell/Modules                       # user level
/usr/local/share/powershell/Modules                                # all users level
/usr/local/microsoft/powershell/6-preview/Modules                  # internal
```

let's say they have version 4.5 in their _user module path_ ( but 4.6 in their _allusers module path_

This function will return true but 4.5 is what will get imported when doing:

`Import-Module Pester` 
or during module auto-loading

I think we should only check the first item in the result. This should be the Pester that would get imported.

Reply:
Is that really the behaviour of PowerShell to not use the latest available version? Can you please point to some docs or code for my own education. How do you know that just looking at the first item is enough? `Get-Module` really needs to be improved otherwise to be more useful/predictive...
If that is really the case, then should we not rather call `(Get-Command Invoke-pester).Version` instead (which has the drawback of being 3 times more expensive for some reason...

Reply:
Yes that's the behavior of PowerShell. I'm unaware of any docs out in the wild (cc @SteveL-MSFT or @joeyaiello maybe?).

The `PSModulePath` acts the same way that the `PATH` environment variable does - it goes down the path looking for the command you want to run, then executes the first one it finds.

A simple test:

```powershell
→ Install-Module Pester -Scope AllUsers -Force

→ Install-Module Pester -Scope CurrentUser -Force -MaximumVersion 4.5

→ Get-Command Invoke-Pester

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Function        Invoke-Pester                                      4.5.0      Pester

→ Get-Module -ListAvailable Pester


    Directory: /Users/tyler/.local/share/powershell/Modules

ModuleType Version    Name                                PSEdition ExportedCommands
---------- -------    ----                                --------- ----------------
Script     4.5.0      Pester                              Desk      {Describe, Context, It, Should…}

    Directory: /usr/local/share/powershell/Modules

ModuleType Version    Name                                PSEdition ExportedCommands
---------- -------    ----                                --------- ----------------
Script     4.6.0      Pester                              Desk      {Describe, Context, It, Should…}

```

Reply:
> How do you know that just looking at the first item is enough?

Because when you do `Get-Module` the first item written to the pipeline is first on the `PSModulePath`. That might not be correct though... `Get-Command` is probably better for this.

Reply:
Hmm, yes, this sounds like `Get-Command` is probably the better option. I guess we win the investment back later when `Invoke-Pester` actually gets called. Personally I would keep it in a non-blocking task though and just have it behave like the old system in the first 2 seconds of opening a Pester file...

Reply:
Yeah... can @rjmholt and @SeeminglyScience help me understand why this should be synchronous?

Reply:
AFAICT the bulk of the Pester code lens implementation is sync (other than LSP message reading/writing).  This code is a bit old so maybe we weren't doing as much async then (??) or it could be that the specific symbol we are looking for results in a "fast enough" AST search.

One issue I have with this approach is that it doesn't take into account that I might have already imported an older version of Pester e.g. `Import-Module Pester -RequiredVersion 4.1.1`.  So we should check `Get-Module Pester` first to see if the module is already loaded before getting the list of Pester module.  

Another thing to consider is this feature seems to be implemented before PS Core was shipped.  That is, Windows PowerShell comes with Pester but not PS Core - right?  But this feature is still enabled resulting in this user exp:
```
Invoke-Pester : The term 'Invoke-Pester' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```
Perhaps we should consider shipping Pester with the extension like we do with PSSA?  That would ensure this feature works on a vanilla PS Core install.  This would ensure we have loaded a minimum version of Pester - like we do for PSSA.  But we still allow folks to install **later** versions of Pester and those would get loaded.  Again, very similiar to how we load PSSA.

Reply:
@rkeithhill you touch on a lot of great points.

1. I think we _should_ make this Async. Lets take the `Lazy<bool>` and turn it into a `Lazy<Task<bool>>` and lazily await the value. Shouldn't be too hard to plumb that. The original caller of this function is async already.

2. 
> we should check `Get-Module Pester` first to see if the module is already loaded before getting the list of Pester module

This makes sense. But `Get-Command Invoke-Pester` should cover this if the module is already imported, right? `Get-Command Invoke-Pester` should be all we need, I think.

3. 

> Perhaps we should consider shipping Pester with the extension like we do with PSSA?

Some people rely on specific version of Pester. PowerShell itself relied on a specific version of Pester for a while in preparation for Core support from Pester.

I don't think we can add Pester into PSES and grab the latest version on the PSModulePath because I don't want to hurt customers by forcing a newer version of Pester on them and all of a sudden their CI and their local are different.

PSSA does fall in that camp, a bit... but PSSA is more for local analysis than anything else, I feel.

Reply:
@TylerLeonhardt I agree with point 2 and 3. Re point to, we could potentially offer to install the latest version of Pester as a 3rd option in the dialog, but I see this as an optional, nice to have feature that can happen after this PR.
Regarding point1: I am not sure if we are on the same page: At the moment, when GetPesterLens gets called (which is when a `.tests.ps1` file gets opened and is therefore as late as possible), then we need to know the version of Pester already and have to wait for the result. When I first opened the PR I ran the check in a background thread (`Task.Run` without `.Result`), which meant that until it came back, it reported `false` (which is what most people did not like). I can change it back to this, but I do not need to make it an `Lazy<Task<bool>>` for that, I'd just remove the blocking call to `.Result`. The reason why I have to use `Task.Run` in the first place is because the recommended Execute APIs on the recommended `PSCommand` class only have async methods, therefore it does not matter at all if the function that we call is already async or not.
If you really wanted true async, then we'd have to fix calling methods up to 2 layers above to make it async top to bottom (which is quite a bit more involved for this change), otherwise you wouldn't get any real benefit from it

Reply:
I made it use `Get-Command` now, which makes the code lens appear a second later due to it being more expensive and I also implemented async-await all the way to the top in a minimum viable way.